From 66ddb45546596ce59aec3491ae4b33d1d0691250 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:22:56 -0400
Subject: [x86] perf/intel: Remove division from the intel_pmu_drain_pebs_nhm() hot path

Message-id: <1408462094-14194-8-git-send-email-jolsa@redhat.com>
Patchwork-id: 87996
O-Subject: [PATCH RHEL7.1 BZ1131394 007/325] perf/x86/intel: Remove division from the intel_pmu_drain_pebs_nhm() hot path
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit eb8417aa703eff5ff43d0275f19b0a8e591d818d
Author: Peter Zijlstra <peterz@infradead.org>
Date: Mon Sep 16 09:23:02 2013 +0200

description
===========
Only do the division in case we have to print the result out in a warning.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/perf_event_intel_ds.c b/arch/x86/kernel/cpu/perf_event_intel_ds.c
index 741068b..e79be19 100644
--- a/arch/x86/kernel/cpu/perf_event_intel_ds.c
+++ b/arch/x86/kernel/cpu/perf_event_intel_ds.c
@@ -938,7 +938,7 @@ static void intel_pmu_drain_pebs_nhm(struct pt_regs *iregs)
  struct perf_event *event = NULL;
  void *at, *top;
  u64 status = 0;
- int bit, n;
+ int bit;
 
  if (!x86_pmu.pebs_active)
   return;
@@ -948,16 +948,16 @@ static void intel_pmu_drain_pebs_nhm(struct pt_regs *iregs)
 
  ds->pebs_index = ds->pebs_buffer_base;
 
- n = (top - at) / x86_pmu.pebs_record_size;
- if (n <= 0)
+ if (unlikely(at > top))
   return;
 
  /*
   * Should not happen, we program the threshold at 1 and do not
   * set a reset value.
   */
- WARN_ONCE(n > x86_pmu.max_pebs_events,
-    "Unexpected number of pebs records %d\n", n);
+ WARN_ONCE(top - at > x86_pmu.max_pebs_events * x86_pmu.pebs_record_size,
+    "Unexpected number of pebs records %ld\n",
+    (top - at) / x86_pmu.pebs_record_size);
 
  for (; at < top; at += x86_pmu.pebs_record_size) {
   struct pebs_record_nhm *p = at;
-- 
1.7.1